Document logging via Monolog (#952)
contao/docs: docs/dev/framework/logging.md

---
title: Logging
description: "Use Monolog and custom log channels in Contao."
---

Contao uses [Monolog][Monolog] and the [Symfony Monolog Bundle][MonologBundle]. 
Monolog provides a standardized and simple, yet flexible and extensible solution
to log messages and process them through a variety of sources.

You can use all features of Monolog and the Monolog Bundle in your Contao application.

## Log Channels
In more complex applications, log channels are used to group related messages and
optionally process them differently.

> A typical Symfony application typically uses channels such as `request` or `security`

Reviewer: ```suggestion
A Symfony application typically uses channels such as `request` or `security`
```

to assign messages to the appropriate parts of the application.

Contao itself uses the `contao` channel for all Contao-specific messages.

{{% notice note %}}
Contao also creates additional log channels for sending messages with a specific context.
For more information, see [Contao channels](#contao-channels-convenience-loggers).
{{% /notice %}}

## Log Contexts & Contao's System Log

> The system log in the Contao backend lists all log entries that have been specifically flagged for Contao.

Reviewer 2: ```suggestion
The system log in the Contao back end lists all log entries that have been specifically flagged for Contao.
```

This flagging is done via a special context that is attached to a particular log entry:

```php
use Contao\CoreBundle\Monolog\ContaoContext;

$logger->info(
    'This message is intended for the Contao system log', 
    ['contao' => new ContaoContext(__METHOD__, ContaoContext::GENERAL)]
);
```

> All messages with this `ContaoContext` are automatically send to the system log

Reviewer: ```suggestion
All messages with this `ContaoContext` are automatically sent to the system log
```
